sql/sql_class.cc

```diff
@@ -8001,9 +8035,25 @@ wait_for_commit::wait_for_prior_commit2(THD *thd)
                   &stage_waiting_for_prior_transaction_to_commit,
                   &old_stage);
   while ((loc_waitee= this->waitee.load(std::memory_order_relaxed)) &&
-         likely(!thd->check_killed(1)))
+         (likely(!thd->check_killed(1)) || force_wait))
     mysql_cond_wait(&COND_wait_commit, &LOCK_wait_commit);
-  if (!loc_waitee)
+
+  /*
+    If this worker has been killed prior to this wait, e.g. in do_gco_wait(),
+    then it should not perform thread cleanup if there are threads which
+    have yet to commit. This is to prevent the cleanup of resources that
+    the prior RGI may need, e.g. its GCO. This is achieved by skipping
+    the unregistration of the waitee, such that each subsequent call to
+    wait_for_prior_commit() will exit early (while maintaining the
+    dependence), thus allowing the final call to
+    thd->wait_for_prior_commit() within finish_event_group() to wait.
+  */
+  if (!loc_waitee
```

Comment:
@knielsen, for your attention, the block intends to cover a `SPECULATE_WAIT`:ers, not what MDEV-31448 description means - I gather my mistake.
A  killed `SPECULATE_WAIT`:er is supposed to exit from here without unlinking from the (gc) `order` parent. In the following path down to `finish_event_group` the link to `waitee` must be preserved.
Your [fixes](https://github.com/MariaDB/server/commit/6ce9c839997c1fc78c2989540dd21155a96fb419#diff-1e7bbf2eedc9ef2e8c81bbbf3dcac4287fa55b8bfb6250daae0578814e506a67) of MDEV-31448 do help to preserve the link which I thought to implement with following the pattern of l.8036 (BASE) that is to set `worker_errror` which avoids `unmark()`.

Reply:
@andrelkin , sorry I did not understand what you mean?

I checked my MDEV-31448 fix against the testcase from Jira, the testcase passes with my fix (and without this change to in wait_for_prior_commit2()).

Is there some other problem that is fixed by these changes, and is there a testcase that I can try? Or did you mean that my fix to MDEV-31448 does solve the problem and is a valid alternative to these changes in wait_for_prior_commit2() ?

The wait_for_prior_commit is a very basic mechanism with a precise semantics. I really don't want it to depend on replication internals like thd->rgi_slave->did_mark_start_commit. So if there's anything else needed than my two commits, I want to know what the problem is so I can come up with a good solution.

I don't understand from the comment in Brandon's original patch or the discussions what these changes in wait_for_prior_commit2() are meant to do, or why they are necessary.

Reply:
MDEV-31448 description test I am still reviewing. What I meant is in just attached test which is about the optimistic mode SPECULATE_WAIT worker that get killed in that waiting. The test asserts on your branch too.

I agree the introduction of ``did_mark_start_commit`  is worth of discussion.
`did_mark_start_commit` helps to recognize the worker's wait is not about to commit but rather about `SPECULATE_WAIT` condition to start. The latter is killable unlike the former when the worker's state fits to your [comments](https://github.com/MariaDB/server/blob/bb-10.5-MDEV-13915/sql/sql_class.cc#L8064)
``` 
 But if a wakeup is already in progress, then we must ignore the kill and not give error 
```
I realize now that the MDEV-31448 fixes are still incomplete to leave out a subclass of the killed which is neither `!thd->rgi_slave->did_mark_start_commit` nor `loc_waitee->wakeup_subsequent_commits_running`. So a waiting-for-prior-commit worker that gets killed while the waitee has not started its commit phase still [unlinks](https://github.com/MariaDB/server/blob/bb-10.5-MDEV-13915/sql/sql_class.cc#L8084) from the commit parent. 
Now I am thinking about common coverage of this and  the `did_mark_start_commit` case. It looks it's just enough 
not to do [`remove_from_list()` etc] (https://github.com/MariaDB/server/blob/bb-10.5-MDEV-13915/sql/sql_class.cc#L8082) here, to defer that either to `retry_event_group`, or to `finish_event_group` calling `wait_for_prior_commit2(... force_wait == true)`.

The `wait_for_prior_commit2()` comments indeed need updating. Let me do that short in a separate commit.

Reply:
To follow up
> I realize now that the MDEV-31448 fixes are still incomplete

indeed my SPECULATE_WAIT version of its test asserts in 13915 branch.
> It looks it's just enough not to do remove_from_list() 

And that must be a correct approach. The link with `waitee` remains intact, to play its order-safety role in `finish_event_group` 's `wait_for_prior_commit(... force_wait == true)`.

I am about pushing a commit to that with an elaborated test covering both the SPECULATIVE and commit-time waits.

Reply:
Hm, valid point that do_gco_wait() also continues to wait after being killed (and also after signalling the error).

And I like much better the approach that wait_for_prior_commit doesn't have logic that refers to rgi etc., just skips the unlinking if killed. It's at least a clear semantics, wait_for commit will unlink only once the prior commit either commits or signals that it aborts, not if interrupted by kill. It's a pity that the return doesn't easily distinguish between prior transaction getting error or this transaction getting kill, but we probably don't really need that (or we could add something if we need it).